---
title: "Quantifying and Visualizing Parallel Evolution in Replicate Experimental Populations"
author: "William R. Shoemaker, Jay T. Lennon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
header-includes:
  - \usepackage{array}
output: 
  pdf_document:
    fig_caption: true
geometry: margin=2.54cm
editor_options: 
  chunk_output_type: console
---

## 1) Background

One of the goals of our research is to determine whether independently evolving populations evolve in similar ways under severe energy-limitation. We have been running our own long-term evolution experiments for several years now and want to be able to quantify the degree of parallel evolution within treatment groups, so we can determine whether different treatment groups are evolutionarily diverged. However, the degree of parallelism is often not quantified in experimental evolution, the degree of expected parallelism depends on the unit of interest (i.e., gene, trait, fitness), and the set of available statistical and visual tools is fairly limited. Recent attempts to quantify parallel evolution focused on the excess number of substitutions in a gene relative to a null model (Tenaillon et al., 2016; Good et al., 2017). This approach is innovative and can likely be applied to pooled sequencing data from any evolution experiment that has been run on a long enough time-scale that a large number of mutation origin-fixation events are likely to have occurred. It is worth noting the theoretical implications of this approach. That is, focusing on parallelism at the gene level implies that genotypic space (i.e., the set of all possible genotypic combinations) has been coarse-grained from the nucleotide to the gene level, where the value at a node is now the number of mutations observed in the set of sites within the gene rather than the presence or absence of a mutation at a single site. If the set of evolutionary paths to a fitness optimum (assuming selection is sufficiently strong, and mutation is sufficiently weak) include sites in different genes that have a negative effect on fitness in combination (site epistasis), replicate populations may be enriched for mutations in different sets of genes. 

To accomplish our goal of quantifying the degree of parallel evolution and visualizing evolutionary trajectories in experimental populations, we are taking a three-step approach:

**1)** Apply and developing a multivariate statistical approach to quantify and visualize evolutionary trajectories from pooled population sequence data. 

**2)** Confirm the validity of our approach through simulated data of evolutionary dynamics.

**3)** Test our approach on an existing high-quality dataset.

In order to determine the degree that independently evolving replicate populations arrive at the same evolutionary outcome, we initially extended the G-score measure presented in Tenaillon et al. (2016) to the level of individual populations (rather than being a measure that aggregates across replicate populations) and applied it to our long-term evolution experiment using multivariate ordination techniques (Shoemaker and Lennon, 2017). 

Since then, a more thorough attempt at quantifying parallel evolution was done in a recent study that examined fine-scale temporal pooled sequencing data from Richard Lenski's Long-term Evolution Experiment (Good et al., 2017). We have since worked to apply the measures and statistics presented in Good et al. to our multivariate approach (2017).

As a brief review, in Good et al. (2017) the authors propose a measure of gene multiplicity to detect evolutionary parallelism at the gene level among replicate populations. The multiplicity for each gene is 

$$m_{i}=n_{i}\cdot \frac{\overline{L}}{L_{i}}$$

where $n_{i}$ is the number of mutations in gene $i$ across all replicate populations, $L_{i}$ is the number of nonsynonymous sites in gene $i$, and $\overline{L}$ is the average value of $L_{i}$ across all genes in the genome. Under the null model that the probability that a gene contains a mutation is simply proportional to the length of the gene ($p_{i}\propto L_{i}$), all genes have the same expected multiplicity $\overline{m}=n_{tot}/n_{genes}$, where $n_{tot}$ is the number of mutations among all replicate populations and $n_{genes}$ is the number of genes in the genome. 

In Good et al. (2017) the authors determine that in nonmutator LTEE populations approximately half of all mutations occurred in genes with $m_{i}\geq 2$, twice as many as expected under the null model. The authors concluded that the null model should be replaced with an alternative where mutations are assigned to each gene with probability

$$p_{i}\propto L_{i}r_{i}$$

where $r_{i}$ is an enrichment factor that is not equal to 1. Under the alternative model the maximum likelihood estimator for the enrichment factor is the ratio of observed and expected multiplicities, $r_{i}=m_{i}/ \overline{m}$ and the net increase relative to the null model across all genes is

$$\Delta \ell=\sum_{i}n_{i}\mathrm{log}\left ( \frac{m_{i}}{\overline{m}} \right )$$

The authors note that the maximum likelihood estimate $r_{i}$ may overfit the data and propose that a more appropriate alternative model would be one that focuses on a subset $I$ of the genes where $r_{i}\neq 1$, while the remaining genes have $r_{i}= 1$. The authors identify this set of genes using a cricial $P$-value, $P^{*}$, for a the False Discovery Rate $\alpha=0.05$ and modify the enrichment factors as follows

$$r_{i}= \begin{Bmatrix}
\frac{m_{i}}{\overline{m}} \left ( \frac{1-\frac{\sum_{i\epsilon I} L_{i} }{\overline{L}n_{genes}}}{1-\frac{\sum_{i\epsilon I}n_{i}}{n_{tot}}} \right )& \mathrm{if}\;  i \;  \epsilon \; I\\ 
1& \mathrm{else}.
\end{Bmatrix}$$

This is an innovative approach that builds off of statistical distributions used to describe parallel evolutionary outcomes. However, this measure pools the mutation data for all replicate populations for each gene. To allow for the comparison between replicate populations so that we can begin to develop statistics to determine whether replicate populations have similar evolutionary trajectories from pooled sequencing, the multiplicity statistics presented in Good et al. (2017) need to be deconstructed to the level of individual populations. To accomplish this goal, we propose a multiplicity measure for the $i$th gene in the $j$th population

$$m_{i,j}=n_{i,j}\cdot \frac{\overline{L}}{L_{i}}$$

with the expected multiplicity in population $j$ of $\overline{m_{j}}=n_{tot,j}/ n_{genes}$, giving a log-likelihood compared to the null model (which is now $r_{i,j}=m_{i,j}/ \overline{m_{j}}$) 

$$\Delta \ell_{j}=\sum_{i}n_{i,j}\mathrm{log}\left ( \frac{m_{i,j}}{\overline{m_{j}}} \right )$$

and the modified enrichment factor 

$$r_{i,j}= \begin{Bmatrix}
\frac{m_{i,j}}{\overline{m_{j}}} \left ( \frac{1-\frac{\sum_{i\epsilon I} L_{i} }{\overline{L}n_{genes}}}{1-\frac{\sum_{i\epsilon I}n_{i,j}}{n_{tot,j}}} \right )& \mathrm{if}\;  i \;  \epsilon \; I\\ 
1& \mathrm{else}.
\end{Bmatrix}$$

Using these modified population level gene enrichment scores and the publically available data presented in Good et al. (2017), we calculate the multiplicity score for each gene within each population at each time point for all genes within set $I$, generating a gene-by-population multiplicity matrix. We then built a Bray-Curtis dissimilarity matrix and used Principal Coordinates Analysis (PCoA) to reduce the dimensionality of the dataset and visualize the evolutionary trajectories of the six nonmutator populations. The code and additional analyses can be found in the [GitHub repository](https://github.com/LennonLab/ParEvol).



```{r, results = 'hide', message = FALSE, warning = FALSE, echo=FALSE}
rm(list=ls())
getwd()
setwd("~/GitHub/ParEvol")
knitr::opts_knit$set(root.dir = '~/GitHub/ParEvol/')

# Load dependencies
# Load dependencies
require("vegan")
require("png")
require("grid")
library("scales")
library("data.table")
library("BiodiversityR")
```



```{r, echo=FALSE}
pop_by_gene <- c("data/Good_et_al/gene_by_pop_delta.txt")
df.good <- read.table(paste(pop_by_gene, collapse = ''), sep = "\t", 
                 header = TRUE, row.names = 1)
# only look at nonmutators (for now)
complete_nonmutator_lines <- c('m5','m6','p1','p2','p4','p5')
complete_mutator_lines <- c('m1','m4','p3')
to_keep <- rownames(df) %like% "m5" + rownames(df) %like% "m6" + 
  rownames(df) %like% "p1" + rownames(df) %like% "p2" + 
  rownames(df) %like% "p4" + rownames(df) %like% "p5"
df.noMut <- df.good[as.logical(df.good),]
df.no0 <- df.noMut[apply(df.noMut[,-1], 1, function(x) !all(x==0)),]
```


## 2) Visualizing evolutionary trajectories through coarse-grained genotypic space


```{r, fig.cap = "The first two axes of a PCoA on the LTEE gene-by-population multiplicity matrix. Each population is plotted as a separate figure. Lighter colored points indicate earlier timepoints while darker timepoints indicate later timepoints. The grey dot indicates the ancestor (i.e., no mutations in any gene) \\label{figurelabel}", message = FALSE, warning = FALSE, echo=FALSE, comment = NA}

df.no0.db <- vegdist(df.no0, method = "bray", upper = TRUE, diag = TRUE)
df.pcoa <- cmdscale(df.no0.db, eig = TRUE, k = 3) 
explainvar1 <- round(df.pcoa$eig[1] / sum(df.pcoa$eig), 3) * 100
explainvar2 <- round(df.pcoa$eig[2] / sum(df.pcoa$eig), 3) * 100

times <- c()
for (x in rownames(df.pcoa$points)){
  time <- strsplit(x, '_')[[1]][2]
  times <- c(times, time)
}
# function to return color gradient for time points
get.times.cols <- function(pcoa.points, times){
  # get colors for times
  times.sorted <- as.character(sort(as.numeric(unique(times))))
  number.times <- length(times.sorted)
  colfunc.nonMut <- colorRampPalette(c("lightgreen", "darkgreen"))
  time.cols.nonMut <- colfunc.nonMut(number.times)
  times.cols.pcoa <- c()
  for (x in rownames(pcoa.points)){
    time <- strsplit(x, '_')[[1]][2]
    pop <- toString(strsplit(x, '_')[[1]][1])
    time.position <- match(time, times.sorted)
    if (pop %in% complete_nonmutator_lines){
      time.color <- time.cols.nonMut[time.position]
    }
    times.cols.pcoa <- c(times.cols.pcoa, time.color)
  }
  return(times.cols.pcoa)
}

pcoa.m5 <- df.pcoa$points[rownames(df.pcoa$points) %like% "m5", ]
pcoa.m6 <- df.pcoa$points[rownames(df.pcoa$points) %like% "m6", ]
pcoa.p1 <- df.pcoa$points[rownames(df.pcoa$points) %like% "p1", ]
pcoa.p2 <- df.pcoa$points[rownames(df.pcoa$points) %like% "p2", ]
pcoa.p4 <- df.pcoa$points[rownames(df.pcoa$points) %like% "p4", ]
pcoa.p5 <- df.pcoa$points[rownames(df.pcoa$points) %like% "p5", ]

m5.cols <- get.times.cols(pcoa.m5, times)
m6.cols <- get.times.cols(pcoa.m6, times)
p1.cols <- get.times.cols(pcoa.p1, times)
p2.cols <- get.times.cols(pcoa.p2, times)
p4.cols <- get.times.cols(pcoa.p4, times)
p5.cols <- get.times.cols(pcoa.p5, times)

# make plot
png(filename = paste(c("figs/pcoa.png"), collapse = ''),
    width = 1200, height = 900, res = 96*2)
par(mfrow = c(2, 3),   
    oma = c(5, 4, 0, 0), # two rows of text at the outer left and bottom margin
    mar = c(1.5, 3.1, 1.1, 0.5), # space for one row of text at ticks and to separate plots
    pty="s") # make the plots square

legend("right", inset=c(-0.22,0), 
       legend=c("Ara+2","Ara+4", "Ara-6", "Ara+1", "Ara-5", "Ara+5"), 
       pch=c(0,2,4,1,3,5), title="Pop.", col="darkolivegreen4")

# plots for each population
pcoa.p2.plot <- plot(pcoa.p2[ ,1], pcoa.p2[ ,2],
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7), 
                     pch = 0, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara+2")
abline(h = 0, v = 0, lty = 3)
points(pcoa.p2[ ,1], pcoa.p2[ ,2],
       pch = 0, cex = 1.5, bg = "gray", col = alpha(m6.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

pcoa.p4.plot <- plot(pcoa.p4[ ,1], pcoa.p4[ ,2], 
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7),
                     pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara+4")
abline(h = 0, v = 0, lty = 3)
points(pcoa.p4[ ,1], pcoa.p4[ ,2],
       pch = 2, cex = 1.5, bg = "gray", col = alpha(m6.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

pcoa.m6.plot <- plot(pcoa.m6[ ,1], pcoa.m6[ ,2],
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7), 
                     pch = 4, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara-6")
abline(h = 0, v = 0, lty = 3)
points(pcoa.m6[ ,1], pcoa.m6[ ,2],
       pch = 4, cex = 1.5, bg = "gray", col = alpha(m6.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

pcoa.p1.plot <- plot(pcoa.p1[ ,1], pcoa.p1[ ,2], 
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7),
                     pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara+1")
abline(h = 0, v = 0, lty = 3)
points(pcoa.p1[ ,1], pcoa.p1[ ,2],
       pch = 1, cex = 1.5, bg = "gray", col = alpha(m6.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

pcoa.m5.plot <- plot(pcoa.m5[ ,1], pcoa.m5[ ,2],
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7), 
                     pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara-5")
abline(h = 0, v = 0, lty = 3)
points(pcoa.m5[ ,1], pcoa.m5[ ,2],
       pch = 3, cex = 1.5, bg = "gray", col = alpha(m5.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

pcoa.p5.plot <- plot(pcoa.p5[ ,1], pcoa.p5[ ,2],
                     xlim = c(-0.7, 0.7), ylim = c(-0.7, 0.7),  
                     pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, 
                     axes = FALSE, xlab = '', ylab ='', main = "Ara+5")
abline(h = 0, v = 0, lty = 3)
points(pcoa.p5[ ,1], pcoa.p5[ ,2],
       pch = 5, cex = 1.5, bg = "gray", col = alpha(m6.cols, 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)

box(lwd = 2)

mtext(paste("PCoA 1 (", explainvar1, "%)", sep = ""), side=1, line=1, 
      cex=1.5, col="black", outer=TRUE)  
mtext(paste("PCoA 2 (", explainvar2, "%)", sep = ""), side=2, line=1, 
      cex=1.5, col="black", outer=TRUE)

dev.off()

# Show Plot
img <- readPNG(paste(c("figs/pcoa.png"), collapse = ''))
grid.raster(img)
```

We see from the ordination (Fig. 1) that Ara+2, Ara+4, and Ara-6 have similar non-linear trends in ordination space. Likewise, Ara+1 and Ara-5 have similar trends. However, Ara+5 shows a divergent trend and we do not have an immediate explanation. To determine the sets of genes that contribute to each of these trajectories as well as how the variation in signatures of parallelism changes over time, we will be adapting appropriate measures presented in Good et al. (eqs. 80 - 87 in the supplement) to account for variation between populations.

We can make this trend clearer by plotting the first two axes as a function of time.

```{r, fig.cap = "The first axis of the PCoA axis plotted against time (generations). \\label{figurelabel}",message = FALSE, warning = FALSE, echo=FALSE, comment = NA}
pcoa.first2 <- as.data.frame(df.pcoa$points[,1:2])
colnames(pcoa.first2)[1] <- 'axis1'
colnames(pcoa.first2)[2] <- 'axis2'
split.rows <- do.call(rbind, strsplit(rownames(pcoa.first2), '_'))
pcoa.first2$pop <- split.rows[,1]
pcoa.first2$time <- split.rows[,2]

get.symbol <- function(pcoa.points){
  pop.symbols <- c()
  for (x in rownames(pcoa.points)){
    pop <- toString(strsplit(x, '_')[[1]][1])
    if (grepl("m5", pop)){
      # square
      pop.symbol <- 0
    } else if ( grepl("m6", pop)) {
      pop.symbol <- 1
      # circle
    } else if ( grepl("p1", pop)) {
      pop.symbol <- 2
      # triangle
    } else if ( grepl("p2", pop)) {
      pop.symbol <- 3
      # addition sign
    } else if ( grepl("p4", pop)) {
      pop.symbol <- 4
      # multiplication sign
    } else if ( grepl("p5", pop)) {
      # diamond
      pop.symbol <- 5
    } 
    pop.symbols <- c(pop.symbols, pop.symbol)
  }
  return(pop.symbols)
}

png(filename = paste(c("figs/time_pcoa1.png"), collapse = ''),
    width = 1200, height = 900, res = 96*2)
#par(mar = c(5, 5, 1, 2) + 0.1)
par(mar = c(5, 5, 1, 5) + 0.1, xpd=TRUE)
plot.pcoa1 <- plot(pcoa.first2$time, pcoa.first2$axis1, 
                  xlim = c(0, 63000), 
                  ylim = c(min(pcoa.first2$axis1), max(pcoa.first2$axis1)),
                  pch = 2, cex = 2.0, type = "n", 
                  cex.lab = 1.5, cex.axis = 1.2, 
                  axes = FALSE, xlab = '', 
                  ylab ='')
points(pcoa.first2$time, pcoa.first2$axis1,cex = 1, 
       bg = "gray",  lwd  = 1, 
       pch = get.symbol(pcoa.first2), col = alpha("darkolivegreen4", 0.8))

axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
mtext("Generations", side=1, line=3.5, cex=1.3, outer=FALSE)
mtext(paste("PCoA 1 (", explainvar1, "%)", sep = ""),
      side=2, line=3.5, cex=1.3, outer=FALSE)
box(lwd = 2)
legend("right", inset=c(-0.22,0), 
       legend=c("Ara+2","Ara+4", "Ara-6", "Ara+1", "Ara-5", "Ara+5"), 
       pch=c(0,2,4,1,3,5), title="Pop.", col="darkolivegreen4")
dev.off()
# Show Plot
img <- readPNG(paste(c("figs/time_pcoa1.png"), collapse = ''))
grid.raster(img) 

```

The trajectory of Ara+5 remains fairly constant over time. This may be because Ara+5 is not accumulating mutations in the set of genes that passed false-discovery rate correction ($I$). We will be examining this in subsequent analyses and will conduct the same PCoA analyses on the gene-by-population multiplicity matrix of all protein encoding genes, rather than just the set of genes within set $I$.

For the remaining five nonmutator populations, the trajectory of the major axis of the ordination over time bears a strikng resemblance to the trajectory of fitness gains. Good and Desai (2015) previously determined that a decelerating fitness trajectory by itself provides little power to distinguish between evolutionary models models. Including the mutation trajectory (i.e., the number of mutations per-clone, ($\partial_{t}\overline{M}(t)$)) provided strong constraints on the set of models that explain the fitness trajectory. Perhaps examining the mutation trajectories of individual genes (i.e., a vector of $\partial_{t}\overline{M_{i}}(t)$ for all $i$ rather than $\partial_{t}\overline{M}(t)$ ) would further constrain the set of models that explain the fitness trajectory.

We can examine the degree divergence in the five nonmutator populations (excluding Ara+5) by plotting the second PCoA axis against the number of generations.


```{r, fig.cap = "The second axis of the PCoA axis plotted against time (generations). \\label{figurelabel}", message = FALSE, warning = FALSE, echo=FALSE}
png(filename = paste(c("figs/time_pcoa2.png"), collapse = ''),
    width = 1200, height = 900, res = 96*2)
par(mar = c(5, 5, 1, 5) + 0.1, xpd=TRUE)
plot.pcoa2 <- plot(pcoa.first2$time, pcoa.first2$axis2, 
                  xlim = c(0, 63000), 
                  ylim = c(min(pcoa.first2$axis2), max(pcoa.first2$axis2)),
                  pch = 2, cex = 2.0, type = "n", 
                  cex.lab = 1.5, cex.axis = 1.2, 
                  axes = FALSE, xlab = '', 
                  ylab ='')
points(pcoa.first2$time, pcoa.first2$axis2,cex = 1, 
       bg = "gray",  lwd  = 1, 
       pch = get.symbol(pcoa.first2), col = alpha("darkolivegreen4", 0.8))

axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
mtext("Generations", side=1, line=3.5, cex=1.3, outer=FALSE)
mtext(paste("PCoA 2 (", explainvar2, "%)", sep = ""),
      side=2, line=3.5, cex=1.3, outer=FALSE)
box(lwd = 2)
legend("right", inset=c(-0.22,0), 
       legend=c("Ara+2","Ara+4", "Ara-6", "Ara+1", "Ara-5", "Ara+5"), 
       pch=c(0,2,4,1,3,5), title="Pop.", col="darkolivegreen4")
dev.off()
# Show Plot
img <- readPNG(paste(c("figs/time_pcoa2.png"), collapse = ''))
grid.raster(img)
```


We can see that on the second PCoA axis the trajectories of Ara+1, Ara-5, and Ara-6 overlap while Ara+2 and Ara+4 are fairly close. Again, the trajectory of Ara+5 remains fairly constant over time. 

We are working to apply time-series clustering techniques on the ordination results to explore how multiple populations can be grouped as a single trajectory.

We can identify the set of genes contributing to our observed patterns by plotting explanatory vectors (i.e., gene coordinates) in ordination space. These coordinates reflect the strength and direction that each gene has on the ordination of the different time points and allow us to identify what genes contribute o divergent trajectories in ordination space.

```{r, fig.cap = "The same PCoA as shown in Fig. 1, but with gene coordinates plotted in ordination space. \\label{figurelabel}", message = FALSE, warning = FALSE, echo=FALSE}
# First we calculate the relative multiplicity score of each gene in each sample
df.no0.rel <- df.no0
  for(i in 1:nrow(df.no0)){
    df.no0.rel[i, ] = df.no0[i, ] / sum(df.no0[i, ])
  }

png(filename = paste(c("figs/pcoa_genes.png"), collapse = ''),
    width = 1200, height = 1200, res = 96*2)
#par(mar = c(6.5, 6, 1.5, 2.5) + 0.1, xpd=TRUE)
par(mar = c(5, 5, 1, 5) + 0.1, xpd=T)
# Initiate Plot
plot(df.pcoa$points[ ,1], df.pcoa$points[ ,2],  xlim = c(-0.7, 0.7), 
     ylim = c(-0.7, 0.7),
     xlab = paste("PCoA 1 (", explainvar1, "%)", sep = ""),
     ylab = paste("PCoA 2 (", explainvar2, "%)", sep = ""),
     pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, 
     cex.axis = 1.2, axes = FALSE)
# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
clip(-0.8,0.8, -0.8,0.8)
abline(h = 0, v = 0, lty = 3)

box(lwd = 2)
points(df.pcoa$points[ ,1], df.pcoa$points[ ,2],
       pch = get.symbol(df.pcoa$points), cex = 1.5, bg = "gray", 
       col = alpha(get.times.cols(df.pcoa$points, times), 0.5), lwd  = 3)
points(0, 0, pch = 16, cex = 2.5, bg = 'gray', col = 'gray', lwd = 1.5)
legend("right", inset=c(-0.22,0), 
       legend=c("Ara+2","Ara+4", "Ara-6", "Ara+1", "Ara-5", "Ara+5"), 
       pch=c(0,2,4,1,3,5), title="Pop.", col="darkolivegreen4")

# Now, we use this information to calculate and add species scores
df.pcoa.spec <- add.spec.scores(df.pcoa, df.no0.rel, method = "pcoa.scores")
spec.no0 <- df.pcoa.spec$cproj[apply(df.pcoa.spec$cproj[,-1], 1, function(x) !all(x==0)),]
spec.no0.clean <- spec.no0[(spec.no0[,1] > 0.01) | (spec.no0[,1] < -0.01),] 
spec.no0.clean <- spec.no0.clean[(spec.no0.clean[,2] > 0.01) | (spec.no0.clean[,2] < -0.01),] 
text(spec.no0.clean[ ,1], spec.no0.clean[ ,2], 
     labels = row.names(spec.no0.clean), col = "black")
dev.off()
img <- readPNG(paste(c("figs/pcoa_genes.png"), collapse = ''))
grid.raster(img)





spe.corr <- add.spec.scores(df.pcoa, df.no0.rel, method = "cor.scores")$cproj
corrcut  <- 0.50       # user defined cutoff
imp.spp  <- spe.corr[abs(spe.corr[, 1]) >= corrcut | abs(spe.corr[, 2]) >= corrcut, ]
imp.spp <- imp.spp[rowSums(is.na(imp.spp))!=ncol(imp.spp), ]

# Permutation Test for Species Abundances Across Axes
fit <- envfit(df.pcoa, df.no0.rel, strata =imp.spp, perm = 999)

```

From this visualization, we can see that the *mokC*, *hokB*, and *hokD* genes are close to the later timepoints for the Ara+2 and Ara+4. Whereas *rplF*, *mreB*, and *arcA* tend to be associated with populations Ara-6, Ara+5, and Ara+6. The genes *mokC*, *hokB*, and *hokD* all seem to be part of the type I toxin-antitoxin system, while *rplF*, *mreB*, and *arcA* appear to be part of different molecular pathways.

We can confirm these visual patterns in a more quantitative way by determining the correlation coefficient of each gene along one or PCoA axes, identifying a correlation-coefficient cutoff, and conducting a permutation test.

## 3) Quantifying variation in evolutionary outcomes

A central question we want to answer from our own evolution experiments is how the degree of similarity between evolving populations changes with time. We can roughly see that the populations reach similar points in Euclidean space over time, but we're only looking at a single ordination axis at a time. 

To quantify the degree of similarity between populations over time across multiple ordination axes, we quantified the Euclidean distance between each population and the median for the first three axes of ordination space for all populations sampled at a given time-point (i.e., the centroid). We then plotted the centroid distance for all populations (green points) and the mean centroid distance (black closed circles) against the number of generations. We chose to exclude the population Ara+5 from this analysis since its trajectory was highly divergent from the other nonmutator populations in the first two axes.


```{r fig.cap = "The Euclidean distance to the centroid of all six nonmutator populations plotted against time. The black circles represent mean centroid distance \\label{figurelabel}", message = FALSE, warning = FALSE, echo=FALSE}
df.no0.nop5 <- df.no0[!rownames(df.no0)  %like% "p5", ]
df.no0.nop5.db <- vegdist(df.no0.nop5, method = "bray", upper = TRUE, diag = TRUE)
df.no0.nop5.groups <- sapply(strsplit(rownames(df.no0.nop5),"_"), `[`, 2)

beta.disp <- betadisper(d = df.no0.nop5.db, group = df.no0.nop5.groups)

get.euc.dist.2D <- function(beta.disp, groups, axes_number){
  centroids <- beta.disp$centroids[,1:axes_number]
  posistions <- beta.disp$vectors[,1:axes_number]
  # get euclidean distancs from first three axes
  eucs <- c()
  pop.name <- c()
  times <- c()
  for(i in groups) {
    centroid.i <- centroids[i, ]
    positions.i <- posistions[rownames(posistions) %like% paste("_", i, sep = ""), ]
    for(j in 1:nrow(positions.i)) {
      position.j <- positions.i[j,]
      sample <- rownames(positions.i)[j]
      euc.dist.j <- dist(rbind(position.j, centroid.i))
      eucs <- c(eucs, euc.dist.j)
      pop.name <- c(pop.name, sample)
      times <- c(times, i)
    }
  }
  pop.euc <- cbind(pop.name, eucs, times) 
  return(data.frame(pop.euc))
}
euc.mat.m <- get.euc.dist.2D(beta.disp, unique(df.no0.nop5.groups), 3)

png(filename = paste(c("figs/mcd.png"), collapse = ''),
    width = 1200, height = 900, res = 96*2)
#par(mar = c(5, 5, 1, 2) + 0.1)
par(mar = c(5, 5, 1, 5) + 0.1, xpd=TRUE)
euc.mat.m$times <- as.integer(as.character(euc.mat.m$times))
euc.mat.m$eucs <- as.numeric(as.character(euc.mat.m$eucs))
rownames(euc.mat.m) <- euc.mat.m$pop.name
euc.mat.m.pch <- get.symbol(euc.mat.m)
x <- euc.mat.m$times 
y <- euc.mat.m$eucs

xy.mean <- aggregate(euc.mat.m[,2], list(euc.mat.m$times), mean )
x.mean <- xy.mean$Group.1
y.mean <- xy.mean$x
plot.mean <- plot(x, y, xlim = c(0, 62000), ylim = c(0, max(y)),
                  pch = 2, cex = 2.0, type = "n", cex.lab = 1.5, 
                  cex.axis = 1.2,  axes = FALSE, xlab = 'Generations', 
                  ylab ='Euclidean distance to centroid')
points(x, y, pch = euc.mat.m.pch, cex = 1, bg = "gray",  
       lwd  = 1, col="darkolivegreen4")
points(x.mean, y.mean, pch = 16, cex = 1.5, bg = "gray", 
       col = alpha('black', 0.6), lwd  = 3)
legend("right", inset=c(-0.22,0), 
       legend=c("Ara+2","Ara+4", "Ara-6", "Ara+1", "Ara-5"), 
       pch=c(0,2,4,1,3), title="Pop.", col="darkolivegreen4")
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
box(lwd = 2)

dev.off()
# Show Plot
img <- readPNG(paste(c("figs/mcd.png"), collapse = ''))
grid.raster(img)

```


The scatter of the points around the mean clearly increases for the first 20,000 generation and then continues to decreases, suggesting that the degree of variance in evolutionary outcomes may be a function of time. 

We have written code to determine whether or not this result is significant by generating random gene-by-population multiplicity matrices where the probability that a mutation lands on the gene is simply the gene length ($p_{i}\propto L_{i}$) and are planning on running a permutation test on the gene-by-population multiplicity matrix.


## 4) Planned analyses

**1)** Extending the historical contingency analysis in Good et al. (2017) to the multivariate case.

**2)** Applying temporal clustering to the PCoA ordination as a means of providing an exploratory analysis of the number of evolutionary trajectories. 

**3)** Determine whether our approach amends itself to time-series analysis by applying certain principles (i.e., temporal autocorrelation) to our analysis.

**4)** Determine if we get comparable results using alternative transformations and ordination techniques (i.e., Hellinger transformed distance matrix and Principal Component Analysis (PCA)).

**5)** Extend our analysis to the polymorphism data, which we can treat as the temporal turnover of genetic diversity (analogous to Beta-diversity).


## 5) Replicate observed patterns via simulation

We are currently working to build off of and simulate existing population genetic models to confirm our observed patterns. We are aiming to take a two-prong approach towards deriving and simulating evolutionary models of parallel evolution:

**1) "Bottom-up"**: For this approach we will simulate genotypic spaces under different evolutionary scenarios (i.e., sign epistasis, historical contingency, etc.), which will then be coarse-grained into individual genes. We've been paying particular attention to the block model (Perelson and Macken, 1995; Schmiegelt and Krug, 2014).

**2) "Top-down"**: Here the degree of parallel evolution is constrained by the genetic background dependence of the distribution of fitness effects. This form of epistasis has been referred to as "macroscopic epistasis" and has been used in a previous approach towards inferring the evolutionary dynamics of the LTEE (Good and Desai, 2015). This approach has the potential advantage of providing a general constraint on the set of evolutionary outcomes based on few pieces of information, since it does not include information about the background dependence of individual mutations. 

Our primary aim is to simulate data from different theoretical approaches under various evolutionary dynamics to test the limits of an ordination approach. We are also interested in analytic derivations to describe the degree that parallel outcomes can be observed by methods that coarse-grain genotypic space to the gene level (i.e., the multiplicity score) and are open to exploring the degree of parallelism under alternative regimes (i.e., clonal interference).


## 6) References

Good, B. H., and M. M. Desai. 2015. The impact of macroscopic epistasis on long-term evolutionary dynamics. Genetics 114:172460 

Good, B. H., M. J. McDonald, J. E. Barrick, R. E. Lenski, and M. M. Desai. 2017. The dynamics of molecular evolution over 60,000 generations. Nature 551:45–50

Perelson, A.S., C. A. Macken. 1995. Protein evolution on partially correlated landscapes. Proc. Natl. Acad. Sci. USA 92:8657–9661

Schmiegelt. B., and J. Krug. 2014. Evolutionary accessibility of modular fitness landscapes. Journal of Statistical Physics 154:334–355

Shoemaker, W. R., and J. T. Lennon. 2017. The contribution of dormancy to microbial evolution. Society for Molecular Biology and Evolution, Austin, Texas, USA

Tenaillon, O., et al. 2016. Tempo and mode of genome evolution in a 50,000-generation experiment. Nature 536:165–170